docs/proposals/container-runtime-interface/capabilities.md

```diff
@@ -0,0 +1,221 @@
+# Linux Capabilities in Container Runtime Interface
+
+The following proposal summarizes the current state and problems for runtimes to implement Linux Capabilities.
+
+## Problems
+
+Today's [Linux Capabilities Interface](https://github.com/kubernetes/kubernetes/blob/v1.5.0-alpha.0/pkg/kubelet/api/v1alpha1/runtime/api.proto#L350-L355) in CRI is derived from the Docker runtime implementation, which also maps to the [k8s API](https://github.com/kubernetes/kubernetes/blob/v1.5.0-alpha.0/pkg/api/v1/types.go#L1164-L1170).
+
+If users want to launch a container with ad-hoc capabilities, they will go through the following steps:
+
+- Get a list of capabilities he would require
+- Look up Docker's docs about the [default capability list](https://docs.docker.com/engine/reference/run/#/runtime-privilege-and-linux-capabilities).
+- Calculate the `Add` and `Drop` capability set based on his requirements and the default capabilities list.
+- Specify the `Add` and `Drop` set for the containers in a k8s pod spec.
+- Launch the pod/container.
+
+There are several issues here:
+
+- The default capability set heavily depends on Docker's implementation; Kubernetes has no control over that.
+When Docker changes the default capability set across versions, the effective set of default capabilities can change suddenly.
+- To implement the same semantics as today's Docker runtime, other runtimes (e.g. rkt) are required to depend on the implicit Docker default capability set as well. If Docker changes the default capability set, then consistency is not guaranteed.
+- From a security point of view, an explicit capability set which does not rely on undefined defaults is preferable for auditing purposes.
+
+## Solution
+
+The proposed solution will:
+
+- Keep the Kubernetes API backwards compatible.
+- Add a new `RequestedSet` field to the `PodSpec.SecurityContext.Capabilities` struct, to allow users to request a specific set of capabilities.
+- Add a new `DefaultCapabilities` field to the `PodSecurityPolicySpec` struct, to allow administrators to define a default set of capabilities.
+- Define a default capability whitelist inside the following Kubernetes components:
+  1. the *API server*: Capability defaults are being applied to `PodSpec.SecurityContext.Capabilities.RequestedSet` either through the `PodSecurityPolicySpec` or, if admission control is disabled, using a default capability whitelist.
+  2. the *Kubelet*: If the kubelet receives a `PodSpec.SecurityContext.Capabilities.RequestedSet` that does not specify any capabilities, the kubelet applies capability defaults to retain backwards compatibility. This could occur for example when a kubelet is newer than the API server.
+- Make CRI expose the whitelist based capability interface, remove the existing `add_capabilities`, and `drop_capabilities` fields, and put the burden on the runtime to implement that.
+- Make the final requested capability set **visible** through the container spec, for example:
```

Comment:
This is fine if the defaulting is applied in the apiserver, but I'm not sure there are any existing examples of the Kubelet modifying the PodSpec. This is probably a non-issue, since the only time that would come into play is if the apiserver was older, in which case it would ignore the update from the Kubelet anyway. (it might be worth adding the caveat here)